### PR TITLE
Hash random bytes before putting them on the wire

### PR DESCRIPTION
**Pull request by Liamsi**
As discussed in #23

### CODE REVIEW

**Comment by c633** [comment]
LGTM.

**Comment by masomel** [comment]
Same, LGTM.

**Comment by arlolra** [comment]
Since we're doing the same ceremony three times, add a helper?

```
func makeRand() []byte, error {
  rand := make([]byte, crypto.HashSizeByte)
  if _, err := rand.Read(rand); err != nil {
    return nil, err
  }
  // Do not directly reveal bytes from rand.Read on the wire
  return crypto.Digest(rand), nil
}
```

**Comment by Liamsi** [comment]
> Since we're doing the same ceremony three times, add a helper?

We are doing it only twice (the exact same ceremony). But that's already enough for creating the helper I think. But we could hash the (already hashed `rand.Read` output from here: https://github.com/coniks-sys/coniks-go/pull/40/commits/929c98c7adc122c8af9779fbaac7483713b01a68#diff-475ca4a45242c730e6134d970fe64530R59) once again to be consistent with the 2 other calls (and hashing is "cheap" enough).

Do you think this should go into the `crypto` package? Currently, it will only be used in the `merkletree` package

**Comment by masomel** [comment]
> Do you think this should go into the ``crypto`` package? Currently, it will only be used in the ``merkletree`` package

The ``computePrivateIndex()`` isn't in ``crypto`` and only used in ``merkletree``. But I think ``makeRand()`` is a bit more general, so I think putting it in ``crypto`` makes sense.

**Comment by arlolra** [comment]
If you revert the changes you made to pad.go, it's three.  But I guess the double hash is cheap enough that it isn't really an issue.

**Comment by Liamsi** [comment]
> If you revert the changes you made to pad.go, it's three. But I guess the double hash is cheap enough that it isn't really an issue.

OK, but it still unnecessary. Reverted it. Thanks @arlolra !

File: merkletree/merkletree.go [hunk 1]
```diff
@@ -26,10 +25,11 @@ type MerkleTree struct {
 
 func NewMerkleTree() (*MerkleTree, error) {
 	root := NewInteriorNode(nil, 0, []bool{})
-	nonce := make([]byte, crypto.HashSizeByte)
-	if _, err := rand.Read(nonce); err != nil {
+	nonce, err := crypto.MakeRand()
+	if err != nil {
 		return nil, err
 	}
+	nonce = crypto.Digest(nonce)
```
Comment (arlolra) [inline merkletree/merkletree.go]: Remove :)

File: merkletree/pad.go [hunk 2]
```diff
@@ -46,17 +45,18 @@ func NewPAD(policies Policies, key crypto.SigningKey, len uint64) (*PAD, error)
 
 // if policies is nil, the previous policies will be used
 func (pad *PAD) signTreeRoot(m *MerkleTree, epoch uint64) {
-	var prevStrHash []byte
+	var prevSTRhash []byte
```
Comment (arlolra) [inline merkletree/pad.go]: Maybe just call it `prevHash` to match the arguments to `NewSTR`.

File: merkletree/merkletree.go (hunk 1 above)
Comment (Liamsi) [inline merkletree/merkletree.go]: Ooops!

File: merkletree/pad.go (hunk 2 above)
Comment (Liamsi) [inline merkletree/pad.go]: Makes sense. Done.

Comment (arlolra) [inline merkletree/pad.go]: Except that you did `prevhash`, all lowercase :)

Comment (Liamsi) [inline merkletree/pad.go]: Damm. Sorry for the noise!

**Comment by arlolra** [comment]
Merged in c091b275d7ce823253b591ed263a50004e025962.

@Liamsi Please see https://github.com/coniks-sys/coniks-go/pull/39#issuecomment-236002541

**Comment by Liamsi** [comment]
> @Liamsi Please see #39 (comment)

Thanks. Will rebase next time.

**Pull request closed without merging**